Slider jumping issue (#807) and infinite scroll realigns rows issue #(761)

juzen2003 opened the pull request:
- I accindently kill https://github.com/SETI/pds-opus/pull/809 when renaming my remote branch, so I recreate this pull request that combines previous two pull requests (since they are modifying the similar function and area).
- This pull request is related to the following issues: #807 and #761 
- This pull request also combines the following two old pull requests: 
  - https://github.com/SETI/pds-opus/pull/809
  - https://github.com/SETI/pds-opus/pull/806
- local branch: slider_issue_807_and_infiniteScroll_issue_761
- Now the scrollbar position will not be re-aligned after infiniteScroll load is triggered.
- Now when scrolling in table view, switch to gallery view, and go back to table view again, the scrollbar position at table view will still be at the same observation number as previous one if the startObs is not changed in gallery view. For example: (row boundary: 15)
  - The current slider number is 8206 in table view. We scroll down to 8218 as the top item in table view. Switch to gallery view, and switch back to table view again. 8218 will still be the top item in table view because the startObs is not changed in gallery view.
  - The current slider number is 11026 in table view. We scroll down to 11032 as the top item in table view. Switch to gallery view and scroll down so that 11041 is the new startObs. Switch back to table view, the top item will be 11041 because the startObs is changed in gallery view and slider is updated to 11041.

juzen2003 committed:
- In browse.js (line 612, 623-624, 634), update setScrollbarPosition
  function to take in one more parameter "offset" (default to 0), this
  will be used to more properly set the scrollbar location so that
  scrolling looks smoother.
- In browse.js (line 757-773, 781-782, 786-787), update updateSliderHandle
  to get the scrollbar offset and store it in infiniteScroll instances as
  "scrollbarOffset". This will be used later on to pass in
  setScrollbarPosition after infiniteScroll load event is triggered and
  renderGalleryAndTable is called.
- In browse.js (line 656-657, 661, 666), update onUpdateSlider to set
  offset to 0 in infiniteScroll instances so that the startObs item can be
  displayed fully after slider has done moving.
- In browse.js (line 1569-1572), update path in infiniteScroll. We will
  update the obsNum in InfiniteScroll instances with the startObs values
  right before infiniteScroll load event is triggered. This will be used
  to properly set the scrollbar and slider location after new data is loaded.
- In browse.js (line 1323), pass in infiniteScroll instances' scrollbarOffset
  to setScrollbarPosition. This will make scrolling looks smoother when
  infiniteScroll load event is triggered.

juzen2003 committed:
- In browse.js (line 714-721), update updateSliderHandle so that when
  calculating startObs in table view, we are using the 2nd table tr's
  height. This is because in some corner cases, the 1st table tr's height
  is 1px larger than rest of tr and it will mess up the calculation.

juzen2003 committed:
- In browse.js (line 753), in table view, we divide by the 2nd table tr's
  height because in some corner cases, the first table tr's height will
  be 1px larger than rest of tr, and this will mess up the calculation.
- In browse.js (line 699-711, 728-854), break down steps in updateSliderHandle
  into multiple functions and add comments.

juzen2003 committed:
- In browse.js (line 1190-1191), update updateBrowseNav function to set
  the scrollbar position using scrollbarObsNum stored in infiniteScroll
  instance in table view.
- In browse.js (line 773-786), modify getCurrentStartObs function to properly
  update currentScrollObsNum so that later on table view scrollbar location
  stays at where it's been left off when we switch back to table view again.

juzen2003 committed:
- In browse.js (line 727), rename "getCurrentStartObs" function to
  "realignDOMAndGetStartObsAndScrollbarObsNum".
- In browse.js (line 743), rename variable "calculatedFirstObs" to
  "alignedCachedFirstObs".
- In browse.js (line 738-740), move "deleteCachedObs" function call inside
  "realignDOMAndGetStartObsAndScrollbarObsNum".
- In browse.js (line 769-771, 774,777), update the comments for
  currentScrollObsNum and obsNum inside "realignDOMAndGetStartObsAndScrollbarObsNum".

juzen2003 committed:
- In browse.js (line 781-795), modify the logic about getting obsNum and
  currentScrollObsNum. Now in gallery view, if scrollbarObsNum in
  infiniteScroll instance is:
  (1) still within the current startObs' boundary
  (2) larger than max slider value when current startObs is equal to
      max slider value
  We want to make sure it didn't get updated to startObs. This will make
  sure table view scrollbar location stays at where it's been left off
  when we switch back to table view again.

juzen2003 committed:
- In browse.js (line 790), adding ;

juzen2003 committed:
Merge branch 'infiniteScroll_issue_761' into slider_issue_807

dstopp commented on opus/application/static_media/js/browse.js:
```diff
@@ -796,6 +731,176 @@ var o_browse = {
+        // For table view, we will set the topBoxBoundary to be the bottom of thead
+        // (account for height of thead)
+        let topBoxBoundary = (o_browse.isGalleryView() ?
+                              $(`${tab} .gallery-contents`).offset().top :
```
might want to consider creating a variable for the gallery-contents   .top value, as it seems to be used in multiple places

rfrenchseti replied:
I think this has not been resolved.

dstopp commented on opus/application/static_media/js/browse.js:
```diff
@@ -796,6 +731,176 @@ var o_browse = {
+        return [obsNum, currentScrollObsNum];
+    },
+
+    deleteCachedObs: function(tab, viewNamespace ,galleryBoundingRect, firstCachedObs) {
```
is this different than the function:
    deleteCachedObservation: function(galleryObsElem, tableObsElem, index, viewNamespace) {
the names are so similar...

rfrenchseti replied:
Agreed. This function needs a better name.

dstopp commented on opus/application/static_media/js/browse.js:
```diff
@@ -696,90 +701,20 @@ var o_browse = {
+            let maxSliderVal = o_browse.getSliderMaxValue(viewNamespace, galleryBoundingRect);
+            let [galleryOffset, tableOffset] = o_browse.getScrollbarOffset(tab, obsNum, currentScrollObsNum);
+
+            o_browse.updateSliderValue(tab, galleryBoundingRect, startObsLabel, maxSliderVal,
```
was curious why you made this a function when it is only used once and it takes 7 input parameters.  I can see that it shortens the original function, is that the reason?  I'm on the fence about that sort of function creation.

rfrenchseti replied:
I haven't looked at the code carefully here, but I think Debby thought that most of these passed parameters could actually be computed inside the called function since they were only used for that one purpose. That would reduce the number of passed arguments and make it look cleaner.

dstopp replied:
I think that the only thing you should actually pass in is obsNum , currentScrollObsNum, galleryOffset & tableOffset.  The var maxSliderVale is only used by this function and can be retrieved in the function; the others can just be retrieved inside the function as it is not time consuming or cycle-intensive to just get those values again. 
In addition, getScrollbarOffset should return an object vs. a list of parameters <see interesting discussion in slack>. This will avoid potential future issues w/parameter ordering, etc.

dstopp commented on opus/application/static_media/js/browse.js:
```diff
@@ -796,6 +731,176 @@ var o_browse = {
+        let galleryOffset = 0;
+        let tableOffset = 0;
+        if (galleryTarget.length && tableTarget.length) {
+           let galleryTargetTopPosition = galleryTarget.offset().top;
```
note that the tab space here is 3 :)  just an observation...

dstopp commented on opus/application/static_media/js/browse.js:
```diff
@@ -1295,7 +1401,7 @@ var o_browse = {
         //   it for the case when cached data is removed so that scrollbar position is always correct (and never reaches to the
         //   end until it reaches to the end of the data)
         o_browse.setScrollbarPosition(infiniteScrollData.options.sliderObsNum,
```
I think you can simplify the code a little bit here if you make infiniteScrollData.options a variable.

dstopp commented on opus/application/static_media/js/browse.js:
```diff
@@ -1541,12 +1647,10 @@ var o_browse = {
+                        // right before infiniteScroll load event is triggered. This will be used to properly
+                        // set the scrollbar and slider location after new data is loaded.
+                        let scrollbarObsNum = opus.prefs[startObsLabel];
                         scrollbarObsNum = Math.max(scrollbarObsNum, 1);
```
can scrollbarObsNum ever be less than 1?

juzen2003 replied:
Because scrollbarObsNum is the top item in table view (or one of the top row items in gallery view), the minimum number will be 1.

rfrenchseti replied:
I think Debby's point was that if opus.prefs[startObsLabel] can never be less than 1, then there's no point in doing the max() and it can be removed.

dstopp commented on opus/application/static_media/js/browse.js:
```diff
@@ -696,90 +701,20 @@ var o_browse = {
-            $(`${tab} .op-slider-nav`).removeClass("op-button-disabled");
-            // update startobs in url when scrolling
-            o_hash.updateHash(true);
+            // When we update the slider, the following steps are run:
```
I think renaming the variable "numThumbnails" to numObservations <or similar> would help the code at line 699 be clearer because 'selector' may refer to the number of thumbnails on the gallery page OR the number of observations in the viewable part of the table,

juzen2003 committed:
- In browse.js (line 699-700), rename "numThumbnails" to "numObservations".
- In browse.js (line 807), rename "deleteCachedObs" to
  "deleteObsToCorrectRowBoundary".
- In browse.js (line 844-848), update updateSliderValue function to take
  less parameters, instead of taking in those parameters, those parameters
  are obtained inside the function.
- In browse.js (line 896-902), fix the indentation.
- In browse.js (line 1305, 1407-1408), simplify the code a little bit by
  making infiniteScrollDataObj = $(selector).data("infiniteScroll").options;

juzen2003 committed:
- In browse.js (line 807, 908), update the return of
  realignDOMAndGetStartObsAndScrollbarObsNum and getScrollbarOffset. Now
  these two functions will return an object instead of a array.

juzen2003 committed:
- In browse.js (line 735, 810, 830, 892, ), modify
  realignDOMAndGetStartObsAndScrollbarObsNum, deleteObsToCorrectRowBoundary,
  getScrollbarOffset and getSliderMaxValue to take in less parameters.
  Parameters like: tab, viewNamespace, and galleryBoundingRect will be
  obtained in each function.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -691,99 +696,30 @@ var o_browse = {
-            o_hash.updateHash(true);
+            // When we update the slider, the following steps are run:
+            // 1. Get the current startObs and top obsNum (top item in table view and one of
+            //    top row items in gallery view) that scrollbar is at.
```
Need to add a comment here about getting the current startObs etc. also doing a DOM realignment, and why that's necessary

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -691,99 +696,30 @@ var o_browse = {
+            let maxSliderVal = o_browse.getSliderMaxValue();
+            let scrollbarOffset = o_browse.getScrollbarOffset(obsNum, currentScrollObsNum);
+            let galleryOffset = scrollbarOffset.galleryOffset;
+            let tableOffset = scrollbarOffset.tableOffset;
```
I don't think there's really much point to having all these extra "let" statements here when the variable is only used once. Just put obsNumObj.scrollbarObsNum or whatever directly in the function call argument list.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -796,6 +732,187 @@ var o_browse = {
     },
 
+    realignDOMAndGetStartObsAndScrollbarObsNum: function(selector, browserResized) {
+        /**
```
Add comment about what we're doing to the DOM

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -796,6 +732,187 @@ var o_browse = {
+    deleteObsToCorrectRowBoundary: function(tab, firstCachedObs) {
+        /**
+         * If browser window is resized, delete some cached observations
+         * to make sure row boundary is correct.
```
This is such a weird thing to do that I think it would be helpful to have a much more descriptive comment so future developers will understand why we're doing this. Say something about the automatic wrapping around of the thumbnails and how when the browser is resized the first element in the cache won't necessarily be on an appropriate row boundary anymore, and we have to delete obs to make that true, because we always want the cache to be on a row boundary.

rfrenchseti reviewed (changes requested):
- Choose a search that only gives you less than one page of results such that the slider is disabled; look at the URL and notice that startobs is negative! Then when you change your search to be larger startobs still stays negative and the slider has a negative value.
- Look at table view. Move the slider somewhere. Note the top line of the table. Resize the browser. The top line of the table will be different. It seems to me like this is unnecessary. Given the control we have over scrollbarObsNum and startObsNum we should be able to put the top row of the table back to the same obs.
- When you resize the browser, the slider jumps to one number, and then quickly to another number.

juzen2003 committed:
- In browse.js (line 757-760), declare browseContentsContainerTop to store
  $(`${tab} .gallery-contents`).offset().top to make the code cleaner.
- In browse.js (716-717), remove declaration of galleryOffset and tableOffset.
  Since they are used for one time only, we pass in the values from the
  object directly.
- In browse.js (line 1664), remove unnecessary line
  scrollbarObsNum = Math.max(scrollbarObsNum, 1);

juzen2003 committed:
- In browse.js (line 705-710), update comments in updateSliderHandle.
- In browse.js (line 740-750), update comments in
  realignDOMAndGetStartObsAndScrollbarObsNum.
- In browse.js (line 823-831), update comments in
  deleteObsToCorrectRowBoundary.

juzen2003 committed:
- In browse.js (line 760-762), modify realignDOMAndGetStartObsAndScrollbarObsNum
  to update firstCachedObs after browser is resized. This will make sure
  slider is updated to the correct value when browser is resized. Previously
  we didn't update firstCachedObs after browser is resized, so the obsNum
  is one row ahead at first. And the correct obsNum gets updated when the
  mutationObserver detects changes in gallery view. That's why we saw the
  slider changes the value twice when window is resized. Now slider will
  change to the correct value right away.

juzen2003 committed:
- Commit before switching branches, leave debugging console.log in browse.js.

juzen2003 committed:
- In browse.js (line 821-835), update the logic in realignDOMAndGetStartObsAndScrollbarObsNum
  so that When resizing happened and some observations are deleted to
  correct the row boundary:
  (1) In table view, if scrollbarObsNum is still within the current startObs'
  boundary, scrollbar will stay at where it is.
  (2) In table view, if scrollbarObsNum is not within the current startObs'
  boundary, scrollbar will be moved to the location where slider value
  (current startObs) is the top item.

juzen2003 committed:
- In browse.js (line 725), update the if condition to (opus.prefs[startObsLabel]
  + numObservations - 1) so that we can get the correct number of rendered
  observations.

juzen2003 committed:
- In browse.js (line 791, 802-803), add defensive programming to make sure
  obsNum is never less than 1.

juzen2003 committed:
- In browse.js (line 640), rename function "onUpdateSliderHandle" to
  "onSliderHandleMoving".
- In browse.js (line 646), rename function "onUpdateSlider" to
  "onSliderHandleStop".
- In browse.js (line 816-853), create a function "calculateCurrentStartObs"
  to include the calculation of obsNum (startObs).

juzen2003 committed:
- In browse.js (line 842-854), update the return of calculateCurrentStartObs
  function.

juzen2003 committed:
- In opus.css (line 110-118), remove min-height for #browse .gallery so
  that there is not extra space for scrolling when #browse .gallery height
  is less than 700px.
- In browse.js (line 738-739), make sure slider is moved to the most left
  when it's disabled.
- In browse.js (line 725-730), update the if condition to make sure when
  slider has no value to scroll, it should always be disabled.
- In browse.js (641, 647, 1693, 1711, 1734), in cart.js (line 249), make
  sure startobs is never less than 1.

juzen2003 committed:
- In browse.js (line 906-907), update getSliderMaxValue to make sure
  return maxSliderVal is never less than 1.
